src/llama-sampling.cpp

```diff
@@ -524,7 +565,223 @@ void llama_sample_top_n_sigma_impl(struct llama_sampling * smpl, llama_token_dat
+        return;
+    }
+
+    // Step 1: Look for restart sequences to limit the maximum repetition length.
+    // Work backwards through the context looking for any token that begins a restart sequence.
+    //
+    // The collection `restart_sequences` is a mapping from a "head" token to all "tail"
+    // sequences that together comprise a restart sequence. This allows us to quickly check
+    // whether each token is the head of a complete sequence. Most restart sequences are actually
+    // a single token, and for these the "tail" is an empty vector.
+    //
+    // If the token is a "head", test all restart sequences that begin with this token
+    // (there will often only be one sequence for each token, but if sequences like 'aaaq1' and
+    // 'aaa1' are used as restart strings, both could start with 'aaa' when tokenized). The
+    // longest matching sequence (if any) is used to limit the maximum repetition length.
+    //
+    // Note that in the case case of a short sequence contained in a longer one, this might fail to
+    // find the smallest value for `rep_limit`. For example, if 'amniotic' and 'ni' are both used as
+    // restart sequences, 'ni' will be found first, and since it's shorter it will fail to suppress
+    // 'otic'. This is a minor issue since fully contained restart sequences are likely to be rare.
+    //
+    // This is theoretically worst-case O(N^2) for arbitrary restart sequences, which is why we
+    // have already clamped the maximum tail sequence length when generating `restart_sequences`.
+    // With clamping, this scan is O(N) in the context length.
+    int32_t effective_dry_penalty_last_n = (dry_penalty_last_n == -1) ?
+        smpl->dry_last_tokens.size() : std::max(dry_penalty_last_n, 0);
+    int last_n_repeat = std::min((int)smpl->dry_last_tokens.size(), effective_dry_penalty_last_n);
+
+    if (last_n_repeat <= dry_allowed_length) {
+        return;
+    }
+
+    // Step 2: Initialize working arrays
+    smpl->dry_repeat_count.assign(last_n_repeat, 0);
+    smpl->dry_max_token_repeat.clear();
+
+    // Step 3: Look for restart sequences (sequence breaker logic)
+    int rep_limit = last_n_repeat;
+    for (int i = 0; i < last_n_repeat; ++i) {
+        llama_token token = smpl->dry_last_tokens[smpl->dry_last_tokens.size() - 1 - i];
+        auto range = smpl->dry_processed_breakers.equal_range(token);
+        if (range.first == smpl->dry_processed_breakers.end()) {
+            continue;
+        }
+        int longest_match = -1;
+        for (auto it = range.first; it != range.second; ++it) {
+            int seq_len = (int)it->second.size();
+            if (seq_len > longest_match && seq_len <= (int)i) {
+                bool match = true;
+                for (int offset = 0; offset < seq_len; ++offset) {
+                    if (it->second[offset] != smpl->dry_last_tokens[smpl->dry_last_tokens.size() - 1 - i + offset + 1]) {
```

Comment:
Is this correct? And even if it is why subtract one then add it?

Reply:
Yes, but that still doesn't answer my question of is it correct? It doesn't look equivalent to the reference implementation to me.